CMSIS/Core/Include/cachel1_armv7.h

```diff
@@ -193,6 +193,25 @@ __STATIC_FORCEINLINE void SCB_DisableDCache (void)
 
     ccsidr = SCB->CCSIDR;
 
+    #if (defined(__GNUC__) && !defined(__OPTIMIZE__))
+      /*
+       * For the endless loop issue with GCC O0.
+       * More details, see https://github.com/ARM-software/CMSIS_5/issues/620
+       *
+       * The issue only happens when local variables are in stack (GCC O0). If
+       * local variables are saved in general purpose register, then the function
+       * is OK.
+       *
+       * When local variables are in stack, after disabling the cache, flush the
+       * local variables cache line for data consistency.
+       */
+      /* Clean and invalidate the local variable cache. */
+      SCB->DCCIMVAC = (uint32_t)(&sets);
+      SCB->DCCIMVAC = (uint32_t)(&ways);
+      SCB->DCCIMVAC = (uint32_t)(&ccsidr);
```

Comment:
Performing three cache line flushes is a bit overkill. It would be better to place the local variables in a struct to ensure co-location, then only perform a cache line flush on that struct's address. Note the necessity to align the struct to the D$ line size.

```c
struct {
    uint32_t ccsidr;
    uint32_t sets;
    uint32_t ways;
} locals __attribute__((aligned(__SCB_DCACHE_LINE_SIZE));

SCB->DCCIMVAC = (uint32_t)&locals;
__DSB();
__ISB();
```

Reply:
Agree, three cache line flushes in a row are definitely an overkill, but it simplifies the rest of the code.